cmd/mdatagen/metrics.tmpl

```diff
@@ -24,36 +24,74 @@ import (
 // Type is the component type name.
 const Type configmodels.Type = "{{ .Name }}"
 
+type metricIntf interface {
+	Name() string
+	New() pdata.Metric
+}
+
+// Intentionally not exposing this so that it is opaque and can change freely.
+type metricImpl struct {
+	name string
+	newFunc func() pdata.Metric
+}
+
+func (m *metricImpl) Name() string {
+	return m.name
+}
+
+func (m *metricImpl) New() pdata.Metric {
+	return m.newFunc()
+}
+
 {{- /* Renders metric descriptors. */}}
-// Metrics contains the metric descriptors for the possible metrics.
+// Metrics contains a set of methods for each metric that help with
+// manipulating those metrics.
 var Metrics = struct {
 	{{- range $name, $metric := .Metrics }}
-	// {{ $name.Render }} in {{ $metric.Unit }} ({{ $metric.Description }})
-	{{ $name.Render }} pdata.Metric
+	{{ $name.Render }} metricIntf
 	{{- end }}
 }{
 	{{- range $name, $metric := .Metrics }}
-	func() pdata.Metric {
-		metric := pdata.NewMetric()
-		metric.InitEmpty()
-		metric.SetName("{{ $name }}")
-		metric.SetDescription("{{ $metric.Description }}")
-		metric.SetUnit("{{ $metric.Unit }}")
-		metric.SetDataType(pdata.MetricDataType{{ $metric.Data.Type }})
-		data := metric.{{ $metric.Data.Type }}()
-		data.InitEmpty()
-		{{- if $metric.Data.HasMonotonic }}
-		data.SetIsMonotonic({{ $metric.Data.Monotonic }})
-		{{- end }}
-		{{- if $metric.Data.HasAggregated }}
-		data.SetAggregationTemporality({{ $metric.Data.Aggregated.Type }})
-		{{- end }}
-
-		return metric
-    }(),
+	&metricImpl{
+		"{{ $name }}",
+		func() pdata.Metric {
+			metric := pdata.NewMetric()
+			metric.InitEmpty()
+			metric.SetName("{{ $name }}")
+			metric.SetDescription("{{ $metric.Description }}")
+			metric.SetUnit("{{ $metric.Unit }}")
+			metric.SetDataType(pdata.MetricDataType{{ $metric.Data.Type }})
+			data := metric.{{ $metric.Data.Type }}()
+			data.InitEmpty()
+			{{- if $metric.Data.HasMonotonic }}
+			data.SetIsMonotonic({{ $metric.Data.Monotonic }})
+			{{- end }}
+			{{- if $metric.Data.HasAggregated }}
+			data.SetAggregationTemporality({{ $metric.Data.Aggregated.Type }})
+			{{- end }}
+
+			return metric
+		},
+	},
 	{{- end }}
 }
 
+var M = Metrics
+
+// MetricsByName makes it easier to look up metrics at runtime by their name.
+var MetricsByName = map[string]metricIntf {
+{{- range $name, $metric := .Metrics }}
+  Metrics.{{ $name.Render }}.Name(): Metrics.{{ $name.Render }},
+{{- end }}
+}
+
+// MetricFactoriesByName makes it easier to get metric factories by name.
+var MetricFactoriesByName = map[string]func() pdata.Metric {
```

Comment:
Is `MetricFactoriesByName` needed?

`MetricsByName["foo"].New()` vs `MetricFactoriesByName["foo"]()` would be better I think to avoid having redundant maps.

Reply:
ie in the future if you want to do `MetricsByName["foo"].Labels()` you'd just have the 1 map instead of having `LabelsByMetricName["foo"]`

Reply:
I forgot to mention that for the Metric helper/builder thing, I am trying to keep the interface between this and that as minimal as possible.  So the most basic and simple thing that needs is a map from metric name to a factory (`func() pdata.Metric`) for it, hence the inclusion.  Having a map from metric name to an `interface{New() pdata.Metric}` is more complex and ties it tighter than necessary to the metadata generator.